src/cli/telemetry/telemetry.ts

```diff
@@ -0,0 +1,203 @@
+import { send } from './ipc';
+import { tryFn, isInteractive } from './helpers';
+import { CompilerSystem, Config, Logger, PackageJsonData } from 'src/declarations';
+import { parseFlags } from '../parse-flags';
+import { loadCoreCompiler } from '../load-compiler';
+import { readJson } from '@ionic/utils-fs';
+import { resolve } from 'path';
+import { readConfig, updateConfig } from '../ionic-config';
+import { isObject } from '@utils';
+
+export interface CommandMetricData {
+	// app_id: string;
+	command: string;
+	arguments: string;
+	options: string;
+	duration: number;
+	error: string | null;
+	node_version: string;
+	// os: string;
+}
+
+export interface Metric<N extends string, D> {
+	name: N;
+	timestamp: string;
+	source: 'stencil_cli';
+	value: D;
+}
+
+// Used for deeper understanding on how people are using certain output targets
+interface OutputTargetOptions {
+	serviceWorker?: boolean;
+}
+
+type TelemetryCallback = (...args: any[]) => void | Promise<void>;
+
+export interface TrackableData {
+	command?: string;
+	token?: string;
+	arguments?: string[];
+	targets?: string[][] | OutputTargetOptions[][];
+	task?: string;
+	duration?: number;
+	packages?: string[][];
+	telemetry?: boolean;
+	ci?: boolean;
+	platform: string;
+	typescript: string;
+	rollup: string;
+	compiler: string;
+	system: string;
+	build: string;
+	stencil: string;
+}
+
+export interface telemetryActionArgs {
+	args?: string[],
+	sys?: CompilerSystem,
+	logger?: Logger,
+	config?: Config,
+	action?: TelemetryCallback
+}
+
+export async function telemetryAction(options: telemetryActionArgs) {
+	const { args, sys, logger, config, action } = options;
+	const coreCompiler = await loadCoreCompiler(sys);
+	const details = sys.details;
+	const versions = coreCompiler.versions;
+	const flags = parseFlags(args, sys);
+	let duration = undefined;
+	let error: any;
+
+	if (action) {
+		const start = new Date();
+
+		try {
+			await action();
+		} catch (e) {
+			error = e;
+		}
+
+		const end = new Date();
+		duration = end.getTime() - start.getTime();
+	}
+	
+	const token = await getTelemetryToken();
+	const packages = await getInstalledPackages();
+	const targets = await getActiveTargets(config);
+
+	const data: TrackableData = {
+		token,
```

Comment:
The service expects the telemetry token to be named `session_id`:
```suggestion
		session_id: token,
```